### PR TITLE
Support listener filter patch

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

implement https://github.com/istio/api/pull/2514

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go [hunk 1]
```diff
@@ -126,9 +126,123 @@ func patchListener(patchContext networking.EnvoyFilter_PatchContext,
 			merge.Merge(listener, lp.Value)
 		}
 	}
+	patchListenerFilters(patchContext, patches[networking.EnvoyFilter_LISTENER_FILTER], listener)
 	patchFilterChains(patchContext, patches, listener)
 }
 
+// patchListenerFilters patches passed in filter if it is MERGE operation.
+// The return value indicates whether the filter has been removed for REMOVE operations.
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go:134]: nit: There is no return here

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go [hunk 2]
```diff
@@ -126,9 +126,123 @@ func patchListener(patchContext networking.EnvoyFilter_PatchContext,
...
+			applied = true
+		} else if lp.Operation == networking.EnvoyFilter_Patch_INSERT_AFTER {
+			// Insert after without a filter match is same as ADD in the end
+			if !hasListenerFilterMatch(lp) {
+				listener.ListenerFilters = append(listener.ListenerFilters, proto.Clone(lp.Value).(*xdslistener.ListenerFilter))
+				applied = true
+				continue
+			}
+
+			// find the matching filter first
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go:161]: Can we refactor this logic in to func? We use it at multiple places to insert a filter at a position?

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch_test.go [hunk 3]
```diff
@@ -893,6 +897,53 @@ func TestApplyListenerPatches(t *testing.T) {
 								"alpn_protocols": [ "h2-6380", "http/1.1-6380" ]}}}}`),
 			},
 		},
+		{
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch_test.go:900]: Can you also please add tests for REMOVE and REPLACE?

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go:161]: I see that, good idea.  Need to dive into it

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go:160]: This exists from beginnging, will clean it in other pr

**Review by ramaraochavali — APPROVED** [review]

**Pull request merged**